### PR TITLE
add support for deep atmosphere

### LINKED ISSUE #2497: Deep atmosphere
@tapios suggested adding support for a deep atmosphere configuration. Here is what I think would need to get done:
1. Update the metric terms so that the horizontal metric terms will scale as 

   $$\frac{z+r}{r}$$

   where $r$ is the radius of the reference sphere, and $z$ is the height from this.

2. Change the Coriolis terms to vary about the planetary rotation axis, basically
    ```julia
    cross(Cartesian3Vector(2 * omega), u)
    ```
    (note this should be a `Cartesian3Vector`, not a `WVector`).
3. We would need to distinguish between vertical line integrals, and vertical "wedge" integrals (which is probably what we would typically want?)

4. Gravity: White (2005) suggests scaling gravitational acceleration vertically by
    $$\frac{(z+r)^2}{r^2}$$
    Note that we should _not_  vary gravity by lat or long, as that would no longer correspond to the gradient of a gravitational potential, which could introduce spurious vorticity

If we did want to account for centrifugal effects in apparent gravity (i.e. lower gravitational acceleration at the equator vs the poles), we could switch to an spheroidal earth. I actually don't think this would be too difficult: main difficulty would be figuring out the appropriate conversions (geodetic lat, long, z) <-> (x1,x2,x3) for computing the metric terms.

Links
- The opposite issue from ClimateMachine.jl: https://github.com/CliMA/ClimateMachine.jl/issues/1270
- White, A.A., Hoskins, B.J., Roulstone, I. and Staniforth, A. (2005), Consistent approximate models of the global atmosphere: shallow, deep, hydrostatic, quasi-hydrostatic and non-hydrostatic. Q.J.R. Meteorol. Soc., 131: 2081-2107. https://doi.org/10.1256/qj.04.49
- Staniforth, A., and N. Wood, 2003: The Deep-Atmosphere Euler Equations in a Generalized Vertical Coordinate. Mon. Wea. Rev., 131, 1931–1938, https://doi.org/10.1175//2564.1. 
- Andrew Staniforth, Nigel Wood, Aspects of the dynamical core of a nonhydrostatic, deep-atmosphere, unified weather and climate-prediction model, Journal of Computational Physics, Volume 227, Issue 7, 2008, https://doi.org/10.1016/j.jcp.2006.11.009.
- Nathan J. Mayne, Isabelle Baraffe, David M. Acreman, Chris Smith, Matthew K. Browning, David Skålid Amundsen, Nigel Wood, John Thuburn and David R. Jackson, The unified model, a fully-compressible, non-hydrostatic, deep atmosphere global circulation model, applied to hot Jupiters - ENDGame for a HD 209458b test case, A&A, 561 (2014) A1, DOI: https://doi.org/10.1051/0004-6361/201322174
- Sebastian Borchert, Guidi Zhou, Michael Baldauf, Hauke Schmidt, Günther Zängl, and Daniel Reinert, 2019, The upper-atmosphere extension of the ICON general circulation model (version: ua-icon-1.0) GMD, 12, 3541–3569, 2019, https://doi.org/10.5194/gmd-12-3541-2019
- Skamarock, W. C., H. Ong, and J. B. Klemp, 2021: A Fully Compressible Nonhydrostatic Deep-Atmosphere Equations Solver for MPAS. Mon. Wea. Rev., 149, 571–583, https://doi.org/10.1175/MWR-D-20-0286.1.

### PR DESCRIPTION
**Pull request by simonbyrne**
## Purpose 
Closes #2497.

The main change is that in a deep atmosphere, the coriolis vector now aligns with the rotational axis, which means the contravariant12 components are no longer zero. I've added the extra terms (which appear naturally on the faces.

It requires https://github.com/CliMA/ClimaCore.jl/pull/1591.

This PR also updates the dependencies to use ClimaCore@0.12, which enables grid construction on GPUs with surface warping. 

## Content

----
- [ ] I have read and checked the items on the review checklist.

### CODE REVIEW

File: src/utils/common_spaces.jl
Comment (simonbyrne) [inline src/utils/common_spaces.jl:84]: This should be a configuration option

**Comment by simonbyrne** [comment]
Tests are failing because they need a new ClimaCore release.

**Comment by simonbyrne** [comment]
I've temporarily added a test Manifest: remove this once a new version is tagged.

**Comment by simonbyrne** [comment]
> I've temporarily added a test Manifest: remove this once a new version is tagged.

Okay, that didn't work. I think you will just have to tag a new ClimaCore first.

**Comment by akshaysridhar** [comment]
Held Suarez: CI diagnostic outcomes with deep-sphere CC + CA branches (55km, 64 vertical faces, day 300)
[summary.pdf](https://github.com/CliMA/ClimaAtmos.jl/files/14117675/summary.pdf)
[summary_sfc.pdf](https://github.com/CliMA/ClimaAtmos.jl/files/14117673/summary_sfc.pdf)

**Comment by szy21** [comment]
@akshaysridhar Thank you. The results look good to me. I agree that we can merge the ClimaCore PR.

**Comment by akshaysridhar** [comment]
Comparing `ClimaAtmos.jl` `e122c4f` (breaking for topography restarts) against `393d2d` (working for topography restarts) with the WIP `ClimaCore.jl` deep-sphere branch.

File: .buildkite/pipeline.yml
Comment (Sbozzolo) [inline .buildkite/pipeline.yml:230]: I added this test to ensure we are using the deep sphere code, but if it doesn't make sense physically, we should remove it and put something else in place.

File: config/model_configs/sphere_baroclinic_wave_rhoe_topography_dcmip_rs_deep.yml
Comment (Sbozzolo) [inline config/model_configs/sphere_baroclinic_wave_rhoe_topography_dcmip_rs_deep.yml:13]: I added this test to ensure we are using the deep sphere code, but if it doesn't make sense physically, we should remove it and put something else in place.

Comment (akshaysridhar) [reply config/model_configs/sphere_baroclinic_wave_rhoe_topography_dcmip_rs_deep.yml:13]: Noted - @Sbozzolo can we tag a CC release? I'm tracking `main` for current tests locally but we should wait for a ClimaCore release since there are major changes to metric reconstruction methods for GPU compatibility.

Comment (Sbozzolo) [reply config/model_configs/sphere_baroclinic_wave_rhoe_topography_dcmip_rs_deep.yml:13]: Yes!

File: Project.toml
Comment (Sbozzolo) [inline Project.toml:52]: Should be `ClimaCore = "0.12"`

**Review by Sbozzolo — APPROVED** [review]

File: config/model_configs/sphere_baroclinic_wave_rhoe_topography_dcmip_rs_deep.yml
Comment (szy21) [reply config/model_configs/sphere_baroclinic_wave_rhoe_topography_dcmip_rs_deep.yml]: Let's add it to this one instead `sphere_held_suarez_rhoe_hightop`. It has a higher top and will show more clearly any effect (or issues) of the deep atmosphere (unless it is intentional to test deep atmosphere with topography? @akshaysridhar). We should add a longrun as well, e.g. this one `longrun_hs_rhoe_dry_55km_nz63`.

File: config/model_configs/deep_sphere_baroclinic_wave_rhoe_equilmoist.yml
Comment (szy21) [inline config/model_configs/deep_sphere_baroclinic_wave_rhoe_equilmoist.yml:1]: We can remove this as we added it to the held-suarez? or do you want to keep both?

Comment (akshaysridhar) [reply config/model_configs/deep_sphere_baroclinic_wave_rhoe_equilmoist.yml:1]: Let's keep both for now. I'd like to move some tests to the GPU longrun list once this is merged, I can clean up the configs at that stage.

**Review by akshaysridhar — APPROVED** [review]

**Pull request merged**
